store/src/nullifier_tree.rs

```diff
@@ -0,0 +1,110 @@
+use miden_objects::{
+    crypto::{
+        hash::rpo::RpoDigest,
+        merkle::{Smt, SmtProof},
+    },
+    notes::Nullifier,
+    Felt, FieldElement, Word, EMPTY_WORD,
+};
+
+use crate::{errors::NullifierTreeError, types::BlockNumber};
+
+/// Nullifier SMT.
+#[derive(Debug, Clone)]
+pub struct NullifierTree(Smt);
+
+impl NullifierTree {
+    /// Construct new nullifier tree from list of items.
+    pub fn with_entries(
+        entries: impl IntoIterator<Item = (Nullifier, BlockNumber)>
+    ) -> Result<Self, NullifierTreeError> {
+        let leaves = entries.into_iter().map(|(nullifier, block_num)| {
+            (nullifier.inner(), Self::block_num_to_leaf_value(block_num))
+        });
+
+        let inner = Smt::with_entries(leaves)?;
+
+        Ok(Self(inner))
+    }
+
+    /// Get SMT root.
+    pub fn root(&self) -> RpoDigest {
+        self.0.root()
+    }
+
+    /// Returns an opening of the leaf associated with the given nullifier.
+    pub fn open(
+        &self,
+        nullifier: &Nullifier,
+    ) -> SmtProof {
+        self.0.open(&nullifier.inner())
+    }
+
+    /// Inserts block number in which nullifier was consumed.
+    pub fn insert(
+        &mut self,
+        nullifier: &Nullifier,
+        block_num: BlockNumber,
+    ) -> Result<(), NullifierTreeError> {
+        let value = self.0.insert(nullifier.inner(), Self::block_num_to_leaf_value(block_num));
+        if value == EMPTY_WORD {
+            return Ok(());
+        }
+
+        Err(NullifierTreeError::NullifierAlreadyExists {
+            nullifier: *nullifier,
+            block_num: Self::leaf_value_to_block_num(value),
+        })
```

Comment:
This error is misleading. It seems to imply the `insert` failed, because the nullifier already exists. But the existing nullifier is being replaced, since `insert` behaves like an `insert_or_replace`.

What was the intention here? To me, it seems that if we want the same behaviour, we should not have this error. OR, if we want to `Err`, then we have to do a check before the insert.

Reply:
Oh! Great catch! Yes, definitely we should do a check and only then do the insert. Basically, if the error is returned, the state of the nullifier tree should not change.